samples/demos/langtons.lobster

```diff
@@ -0,0 +1,311 @@
+// All lobsters that fail their bid turn in place instead.
+
+import gui
+import texture
+
+def pick(list):
+    return list[rnd(length(list))]
+
+// scatter copies the values from the source array into their respective places
+// in an array of all the neighborhoods around their position.
+// The neighborhoods array must contain an array with the indices of each neighbor.
+// So, for an array representing a 3x3 grid in row-major-order:
+//   [ 0, 1, 2,
+//     3, 4, 5,
+//     6, 7, 8 ]
+// the neighborhoods of indices in cardinal directions [north, west, south, east]
+// would be:
+//   [ [ 6, 2, 3, 1 ], [ 7, 0, 4, 2], ... ]
+def scatter(dst, src, neighborhoods):
+    for(neighborhoods) neighborhood, i:
+        for(neighborhood) j, k:
+            dst[i][k] = src[j]
+
+// gather_bids visits each cell and determines the direction the lobster in
+// that cell would like to move, based on its heading.
+// headings are indexes into cardinal_directions, for the direction each
+// lobster is currently heading.
+def gather_bids(bids, neighborhoods, lobsters, headings):
+    for(neighborhoods) neighbors, i:
+        bids[i] = if lobsters[i]:
+            let heading = headings[i]
+            let neighbor = neighbors[heading]
+            if !lobsters[neighbor]: neighbor
+            else: -1
+        else: -1
+
+// The tick method then scatters bids and each lobster’s priority into
+// bid_neighborhoods and weight_neighborhoods.
+
+// gather_winners (through find_winner) examines the bids in each neighborhood
+// and records the bidder with the highest weight to move onto each cell.
+def gather_winners(area, winners, neighborhoods, bid_neighborhoods, weight_neighborhoods):
+    for(area) i:
+        winners[i] = find_winner(i, neighborhoods[i], bid_neighborhoods[i], weight_neighborhoods[i])
+
+def find_winner(i, neighborhood, bid_neighborhood, weight_neighborhood):
+    var best = -1
+    var winner = -1
+    for(length(neighborhood)) j:
+        if bid_neighborhood[j] == i and (best < 0 or weight_neighborhood[j] > best):
+            best = weight_neighborhood[j]
+            winner = neighborhood[j]
+    return winner
+
+// The tick method then builds a transition matrix.
+// The transition matrix contains the index of every cell in the source array,
+// ordered according to where its data were stored in the previous generation.
+
+// transit applies the results of a transition, effectively swapping every
+// cell’s data with that data in its prior location, from the previous
+// generation.
+// The tick function uses transit to move all mobile data from one generation
+// to the next.
+def transit(len, dst, src, transitions):
+    for(len) i: dst[i] = src[transitions[i]]
+
+// move copies one array to another, given the length of both.
+// This is necessary at the beginning of each tick, to overwrite columns of
+// data from the previous generation.
+def move(len, dst, src):
+    for(len) i:
+        dst[i] = src[i]
```

Comment:
Would be neat to have a memcopy builtin for this case. Maybe overload copy, for the no-allocate case. I assume `min(length(src), length(dst))` would be cheap, but I have that figure laying around in this case.

Reply:
Yup that be nice to have